File: yidam/prelude/sdks/rust/src/graph.rs
use std::collections::{HashSet, VecDeque};

#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct GraphEdge {
    pub from: String,
    pub to: String,
}

/// All nodes reachable from `node_path` following directed edges (BFS).
/// The start node is not included. Result is sorted for determinism.
pub fn find_reachable(edges: &[GraphEdge], node_path: &str) -> Vec<String> {
    let mut visited: HashSet<String> = HashSet::new();
    let mut queue: VecDeque<String> = VecDeque::new();
    visited.insert(node_path.to_string());
    queue.push_back(node_path.to_string());
    let mut reachable: Vec<String> = Vec::new();
    while let Some(current) = queue.pop_front() {
        for edge in edges {
            if edge.from == current && !visited.contains(&edge.to) {
                visited.insert(edge.to.clone());
                reachable.push(edge.to.clone());
                queue.push_back(edge.to.clone());
            }
        }
    }
    reachable.sort();
    reachable
}

/// All nodes that have a directed edge pointing to `node_path`.
/// Result is sorted and deduplicated for determinism.
pub fn find_citations(edges: &[GraphEdge], node_path: &str) -> Vec<String> {
    let mut citations: Vec<String> = edges
        .iter()
        .filter(|e| e.to == node_path)
        .map(|e| e.from.clone())
        .collect();
    citations.sort();
    citations.dedup();
    citations
}

#[cfg(test)]
mod tests {
    use super::*;

    fn e(from: &str, to: &str) -> GraphEdge {
        GraphEdge { from: from.into(), to: to.into() }
    }

    #[test]
    fn reachable_linear_chain() {
        let edges = vec![e("a.md", "b.md"), e("b.md", "c.md")];
        assert_eq!(find_reachable(&edges, "a.md"), vec!["b.md", "c.md"]);
    }

    #[test]
    fn reachable_cycle_terminates() {
        let edges = vec![e("a.md", "b.md"), e("b.md", "a.md")];
        assert_eq!(find_reachable(&edges, "a.md"), vec!["b.md"]);
    }

    #[test]
    fn reachable_isolated_node() {
        let edges = vec![e("a.md", "b.md")];
        assert_eq!(find_reachable(&edges, "b.md"), Vec::<String>::new());
    }

    #[test]
    fn citations_single() {
        let edges = vec![e("a.md", "b.md"), e("b.md", "c.md")];
        assert_eq!(find_citations(&edges, "b.md"), vec!["a.md"]);
    }

    #[test]
    fn citations_multiple() {
        let edges = vec![e("a.md", "c.md"), e("b.md", "c.md")];
        assert_eq!(find_citations(&edges, "c.md"), vec!["a.md", "b.md"]);
    }

    #[test]
    fn citations_none() {
        let edges = vec![e("a.md", "b.md")];
        assert_eq!(find_citations(&edges, "a.md"), Vec::<String>::new());
    }
}

File: yidam/prelude/sdks/typescript/src/graph.ts
/** A directed edge between two nodes, identified by corpus path. */
export interface GraphEdge {
  from: string
  to: string
}

/**
 * Code-point order, which `Array.prototype.sort` is not.
 *
 * The default comparator orders by UTF-16 code unit, and an astral character is a surrogate
 * pair beginning in D800–DBFF — below every BMP character from E000 up. Rust sorts `String`
 * by UTF-8 bytes and Python by code point, and those two agree with each other everywhere,
 * so JavaScript is the one runtime that would answer a sorted contract differently. It takes
 * a path pairing an emoji with halfwidth katakana to see it, which is what
 * `astral-and-bmp.toml` is; every ASCII fixture passes with either comparator, and that is
 * exactly the shape a divergence hides in.
 */
function byCodePoint(a: string, b: string): number {
  const ac = Array.from(a)
  const bc = Array.from(b)
  const n = Math.min(ac.length, bc.length)
  for (let i = 0; i < n; i++) {
    const d = (ac[i].codePointAt(0) as number) - (bc[i].codePointAt(0) as number)
    if (d !== 0) return d
  }
  return ac.length - bc.length
}

/**
 * All nodes reachable from `nodePath` following directed edges (BFS).
 * The start node is not included. Result is sorted for determinism.
 */
export function findReachable(edges: GraphEdge[], nodePath: string): string[] {
  const visited = new Set<string>([nodePath])
  const queue: string[] = [nodePath]
  const reachable: string[] = []
  while (queue.length > 0) {
    const current = queue.shift() as string
    for (const edge of edges) {
      if (edge.from === current && !visited.has(edge.to)) {
        visited.add(edge.to)
        reachable.push(edge.to)
        queue.push(edge.to)
      }
    }
  }
  return reachable.sort(byCodePoint)
}

/**
 * All nodes that have a directed edge pointing to `nodePath`.
 * Result is sorted and deduplicated for determinism.
 */
export function findCitations(edges: GraphEdge[], nodePath: string): string[] {
  const citations = edges.filter(e => e.to === nodePath).map(e => e.from)
  citations.sort(byCodePoint)
  return citations.filter((c, i) => i === 0 || c !== citations[i - 1])
}
